docs(entity): add recipe for entity adapters with extra selectors (#3912)
ngrx/platform: projects/ngrx.io/content/guide/entity/recipes/entity-adapter-with-feature-creator.md

# Using Entity Adapter with Feature Creator

> This example demonstrates a User state management using NgRx Store with `Entity Adapters` and `Extra Selectors` for a feature.

Reviewer: Thoughts? It's mainly to add links to the adapter and selector docs.

```suggestion
This recipe demonstrates how to combine [entity adapter](/guide/entity/adapter#entity-adapter) (selectors) within [extraSelectors](/guide/store/feature-creators#providing-extra-selectors).
As an example, let's look at some code from a User state management feature.
```

Usage:

Start by defining the state for the `User` feature.

<code-example header="users.state.ts">

import { EntityState, createEntityAdapter } from '@ngrx/entity';
import { User } from './user.model';

export interface State extends EntityState&lt;User&lt; {
  selectedUserId: string | null;
}

const adapter = createEntityAdapter&lt;User&lt;();

export const initialState: State = adapter.getInitialState({
  selectedUserId: null,
});

</code-example>

Then, we define the `User` actions: `addUser` and `selectUser`.

<code-example header="user.actions.ts">

import { createActionGroup, props } from '@ngrx/store';
import { User } from './user.model';

export const UserListPageActions = createActionGroup({
  source: 'User List Page',
  events: {
    addUser: props&lt;{ user: User }&gt;(),
    selectUser: props&lt;{ userId: string }&gt;(),
  },
});

</code-example>

> Then use the `createReducer` function from NgRx to create a reducer for the `User` feature.

Reviewer: ```suggestion
Then use the `createReducer` function to create a reducer for the `User` feature.

```

<code-example header="users.state.ts">

import { createReducer, on } from '@ngrx/store';

> import { UserListPageActions } from './user-list-page.actions';
> const reducer = createReducer(

Reviewer 2: We can also add an empty line between imports and reducer. The rest of the code snippets look great! :)

```suggestion
import { UserListPageActions } from './user-list-page.actions';

const reducer = createReducer(

```

  initialState,
  on(UserListPageActions.addUser, (state, { user }) =>
    adapter.addOne(user, state)
  ),
  on(UserListPageActions.selectUser, (state, { userId }) => ({
    ...state,
    selectedUserId: userId,
  }))
);

</code-example>

> Then create the `User` feature using the `createFeature` function. This will generate all basic selectors automatically, and we can specify extra selectors using the `extraSelectors` option.

Reviewer: ```suggestion
Then create the `User` feature using the `createFeature` function. 
This generates all the basic selectors automatically, and we can specify extra selectors using the `extraSelectors` option.

```

<code-example header="users.state.ts">

import { createFeature, createSelector } from '@ngrx/store';

export const usersFeature = createFeature({
  name: 'users',
  reducer,
  extraSelectors: ({ selectUsersState, selectEntities, selectSelectedUserId }) => ({
    ...adapter.getSelectors(selectUsersState),
    selectIsUserSelected: createSelector(
      selectSelectedUserId,
      (selectedId) => selectedId !== null
    ),
    selectSelectedUser: createSelector(
      selectSelectedUserId,
      selectEntities,
      (selectedId, entities) => selectedId ? entities[selectedId] : null
    ),
  }),
});

</code-example>

> In the component, we `inject` the `Store` service from NgRx and select the data we need from the state using the selectors generated by the `createFeature` function.

Reviewer: ```suggestion
To use the selector within a component, `inject` the `Store` and select the data from the state using the selectors generated by the `createFeature` function.

```

Reviewer: Rewording this provide a smoother transition on what the next snippet does.
